spec/src/main/asciidoc/variable-expansion.asciidoc

```diff
@@ -0,0 +1,137 @@
+//
+// Copyright (c) 2016-2019 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[variable-expansion]]
+== Variable Expansion
+
+The value of a configuration property can contains a variable corresponding to another configuration property.
+This variable is _expanded_ to be replaced by its own value when the original property is returned.
+
+For example, let's define a `server.url` property:
+
+[source]
+----
+server.url=http://${server.host}:{server.port}/endpoint
+----
+
+When the Config API returns the value of `server.url`, it expands any variable (such as `server.host` and `server.port`) to replace
+them with their value.
+
+If these properties are also defined:
+
+[source]
+----
+server.host=example.org
+server.port=8080
+----
+
+The `server.url` value returned by the Config API is `http://example.org:8080/endpoint`.
+
+[NOTE]
+====
+Backwards Compatibility
+
+Variable expansion is not backwards compatible.
+Previous version of MicroProfile Config would not expand the variables and would return the _raw_ property `http://${server.host}:{server.port}/endpoint`.
+
+Implementation of MicroProfile Config MUST provide a way to disable variable expansion to provide backwards compatibility.
+
+The property `mp-config.expand-variables` can be configured as a boolean in one of the 3 default `ConfigSource` (environment variables, system properties,
```

Comment:
TODO: rename it to `mp.config.expandVariables` to be consistent with other MP specs

Reply:
Could this be implemented as a default method on `ConfigSource`, whose default implementation is to evaluate that property?  This would allow for config source implementations that always or never expand properties.  For example, we've already run into trouble with property expansion in environment variables which contain `${..}` sequences.

Reply:
The intent of this property is a *global* switch that applies at the Config level (either try to expand all properties or not).

If I follow you, you want to be able to configure variable expansion on a `ConfigSource` level.
E.g.

```
# config_source_1
mp.config.source.expandVariables=true
host=localhost
port=8080
server=http://${host}:${port}

# config_source_2
mp.config.source.expandVariables=false
server=http://${host}:${port}
```

If `config_source_1` has higher ordinal, `config.getValue("server", String.class)` would return `http://localhost:8080` (as its variables are expanded).
If `config_source_2` has higher ordinal, `config.getValue("server", String.class)` would return `http://${host}:${port}` (as its variables are *not* expanded).

@dmlloyd is that the behaviour you are expecting?

Reply:
No, I had just misinterpreted the description of the functionality.  I thought the property applied to each config source but the specification states that it must *be configured in* one of the default config sources to apply (globally).

That said, we did internally talk about the fact that we probably do not want to allow expansion within environment variable values as this could cause issues with certain environments.

Reply:
> The intent of this property is a _global_ switch that applies at the Config level (either try to expand all properties or not).
> 
> If I follow you, you want to be able to configure variable expansion on a `ConfigSource` level.
> E.g.
> 
> ```
> # config_source_1
> mp.config.source.expandVariables=true
> host=localhost
> port=8080
> server=http://${host}:${port}
> 
> # config_source_2
> mp.config.source.expandVariables=false
> server=http://${host}:${port}
> ```
> 
> If `config_source_1` has higher ordinal, `config.getValue("server", String.class)` would return `http://localhost:8080` (as its variables are expanded).
> If `config_source_2` has higher ordinal, `config.getValue("server", String.class)` would return `http://${host}:${port}` (as its variables are _not_ expanded).
> 
> @dmlloyd is that the behaviour you are expecting?

Is this implemented in already and can you please let us know how to use it and details.?